### PR TITLE
Run evaluation at end of training

### PR DESCRIPTION
**Pull request by awaelchli**
Runs a final validation loop at end of training, over the entire validation set.
Also refactors out the generated sample into a separate function, so we can run the validation loop without it at the end.

### CODE REVIEW

File: litgpt/finetune/full.py [hunk 1]
```diff
@@ -150,6 +150,14 @@ def main(
     if fabric.device.type == "cuda":
         fabric.print(f"Memory used: {torch.cuda.max_memory_allocated() / 1e9:.02f} GB")
 
+    # Final evaluation
+    val_loss = validate(
+        fabric, model, val_dataloader, tokenizer, dataclasses.replace(eval, max_iters=len(val_dataloader)), data
+    )
+    metrics = {"val_loss": val_loss, "val_ppl": math.exp(val_loss)}
+    fabric.log_dict(metrics, step=state["iter_num"])
+    fabric.print(f"Final evaluation | val loss: {val_loss.item():.4f} | val ppl: {math.exp(val_loss):.4f}")
```
Comment (rasbt) [inline litgpt/finetune/full.py:159]: I wonder what you think about computing this for the training dataset as well? I usually find it useful to look at validation and training performance side by side. It would be expensive though, so maybe a rolling average over the recent training set losses. Could be a separate PR.

**Comment by rasbt** [comment]
That's very nice. I tried to do something similar in #1228 but I like your approach much better. I suggest closing #1228 in favor of your PR.

The other thing is, what do you think about computing the initial validation set loss, too? The reason is 

1) some users got confused about the `n/a` we currently print.

2) it's useful to know how much the model improved from the beginning.

File: litgpt/finetune/lora.py [hunk 2]
```diff
@@ -208,6 +208,14 @@ def main(
     if fabric.device.type == "cuda":
         fabric.print(f"Memory used: {torch.cuda.max_memory_allocated() / 1e9:.02f} GB")
 
+    # Final evaluation
+    val_loss = validate(
+        fabric, model, val_dataloader, tokenizer, dataclasses.replace(eval, max_iters=len(val_dataloader)), data
```
Comment (carmocca) [inline litgpt/finetune/lora.py:213]: Why do you need the `.replace()` call?

Comment (awaelchli) [reply litgpt/finetune/lora.py:213]: Because I want to run on the entire dataset, not just the `eval.max_iters` which for the final validation in my opinion is appropriate here.

Comment (rasbt) [reply litgpt/finetune/lora.py:213]: From a practical standpoint this makes total sense. It's basically the "true" (vs the approximated) validation set performance. I wonder if we can communicate that somehow without making it too confusing. Maybe just by modifying the docstring for eval.max_iters

**Comment by awaelchli** [comment]
@rasbt I wasn't aware of your PR. It's doing a lot of different things that could be discussed separately. For this one I just needed the final validation loss to be able to put it into the benchmark table.

**Comment by rasbt** [comment]
> @rasbt I wasn't aware of your PR. It's doing a lot of different things that could be discussed separately. For this one I just needed the final validation loss to be able to put it into the benchmark table.

No worries, you can ignore the other PR. I can rebase it based on this PR later.

**Comment by rasbt** [comment]
Since the newly added lines are not very wide, should we spell out `ppl` as `perplexity`? It might be less ambiguous to a general audience.

```
poch 1 | iter 74 step 4 | loss train: 1.552, val: n/a | iter time: 134.38 ms
Epoch 1 | iter 75 step 4 | loss train: 1.623, val: n/a | iter time: 134.42 ms
Epoch 1 | iter 76 step 4 | loss train: 1.517, val: n/a | iter time: 150.71 ms
Epoch 1 | iter 77 step 4 | loss train: 1.562, val: n/a | iter time: 148.68 ms
Epoch 1 | iter 78 step 4 | loss train: 1.546, val: n/a | iter time: 161.03 ms
Epoch 1 | iter 79 step 4 | loss train: 1.655, val: n/a | iter time: 154.14 ms
Epoch 1 | iter 80 step 5 | loss train: 1.623, val: n/a | iter time: 165.89 ms (step)
Training time: 15.71s
Memory used: 19.36 GB
Validating ...
Final evaluation | val loss: 1.6684 | val perplexity: 5.3034
```

File: litgpt/finetune/adapter_v2.py [hunk 3]
```diff
@@ -178,6 +178,12 @@ def main(
     if fabric.device.type == "cuda":
         fabric.print(f"Memory used: {torch.cuda.max_memory_allocated() / 1e9:.02f} GB")
 
+    # Final evaluation
+    val_loss = validate(fabric, model, val_dataloader, dataclasses.replace(eval, max_iters=len(val_dataloader)))
+    metrics = {"val_loss": val_loss, "val_ppl": math.exp(val_loss)}
+    fabric.log_dict(metrics)
+    fabric.print(f"Final evaluation | val loss: {val_loss.item():.4f} | val ppl: {math.exp(val_loss):.4f}")
```
Comment (rasbt) [inline litgpt/finetune/adapter_v2.py:185]:
```suggestion
    fabric.print(f"Final evaluation | val loss: {val_loss.item():.4f} | val perplexity: {math.exp(val_loss):.4f}")
```

File: litgpt/finetune/adapter.py [hunk 4]
```diff
@@ -178,6 +178,12 @@ def main(
     if fabric.device.type == "cuda":
         fabric.print(f"Memory used: {torch.cuda.max_memory_allocated() / 1e9:.02f} GB")
 
+    # Final evaluation
+    val_loss = validate(fabric, model, val_dataloader, dataclasses.replace(eval, max_iters=len(val_dataloader)))
+    metrics = {"val_loss": val_loss, "val_ppl": math.exp(val_loss)}
+    fabric.log_dict(metrics)
+    fabric.print(f"Final evaluation | val loss: {val_loss.item():.4f} | val ppl: {math.exp(val_loss):.4f}")
```
Comment (rasbt) [inline litgpt/finetune/adapter.py:185]: Just a suggestion, but to make it consistent with the formatting above, would 3 numbers after the decimal point be sufficient?

```suggestion
    fabric.print(f"Final evaluation | val loss: {val_loss.item():.3f} | val ppl: {math.exp(val_loss):.3f}")
```

File: litgpt/finetune/adapter_v2.py (hunk 3 above)
Comment (awaelchli) [reply litgpt/finetune/adapter_v2.py]: It's not what we log though. I'd like to keep the PR focused on doing one thing only. So if it needs to be renamed, let's do it separately.

Comment (rasbt) [reply litgpt/finetune/adapter_v2.py]: ok that's fine

**Review by rasbt — APPROVED** [review]
